man/amlabel.8
'\" t
.\"     Title: amlabel
.\"    Author: James da Silva <jds@amanda.org>
.\" Generator: DocBook XSL Stylesheets v1.76.1 <http://docbook.sf.net/>
.\"      Date: 07/25/2012
.\"    Manual: System Administration Commands
.\"    Source: Amanda 3.3.2
.\"  Language: English
.\"
.TH "AMLABEL" "8" "07/25/2012" "Amanda 3\&.3\&.2" "System Administration Commands"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
amlabel \- label an Amanda tape
.SH "SYNOPSIS"
.HP \w'\fBamlabel\fR\ 'u
\fBamlabel\fR [\fB\-\-barcode\ \fR\fB\fIbarcode\fR\fR] [\fB\-\-meta\ \fR\fB\fImeta\-label\fR\fR] [\fB\-\-assign\fR] [\fB\-f\fR] [\fB\-\-version\fR] [\fB\-o\fR\ \fIconfigoption\fR...] \fIconfig\fR [\fIlabel\fR] [slot\ \fIslot\fR]
.SH "DESCRIPTION"
.PP
This command applies a label to an Amanda volume, erasing the volume in the process\&.
.PP
Amanda uses labels to identify volumes:
\fBamdump\fR(8)
and
\fBamflush\fR(8)
verify the label before writing to make sure the proper volume is loaded, and the recovery programs use the label to ensure they are reading the data they expect\&.
.PP
If no
[\fIslot\fR]
and no
[\fIbarcode\fR]
is given, then
\fBamlabel\fR
labels the tape in the current slot of the default changer\&. If a slot is given, it labels the volume in that slot\&. If a barcode is given, it labels the volume with that barcode\&.
\fILabel\fR
may be any string that does not contain whitespace and that matches the
\fBlabelstr\fR
(see
\fBamanda.conf\fR(5))\&.
.PP
If
[\fIlabel\fR]
is not given, an autolabel is generated, see
\fBautolabel\fR
in
\fBamanda.conf\fR(5)\&.
.PP
If
[\fB\-\-assign\fR]
is given, the barcode and meta\-label are assigned to the label without labeling the volume\&. The label must already be in the tapelist file\&.
.PP
This command also appends the new volume to the
\fBtapelist\fR(5)
file, so that they will be used in the order they are labeled (depending on the taperscan in use \(em see
\fBamanda-taperscan\fR(7))\&.
.PP
As a precaution,
\fBamlabel\fR
will not write a label if the volume already contains an active label or if the label specified is on an active tape\&. The
[\-f]
(force) flag bypasses these verifications\&.
.SH "OPTIONS"
.PP
\fB\-\-barcode \fR\fB\fIbarcode\fR\fR
.RS 4
With
[\-\-assign], assign the barcode to the label\&. Without
[\-\-assign], label the volume with that barcode\&.
.RE
.PP
\fB\-\-meta \fR\fB\fImeta\-label\fR\fR
.RS 4
Assign the meta label to the label after labeling the volume\&.
.RE
.PP
\fB\-\-assign\fR
.RS 4
Assign the barcode or meta\-label to the label without labeling the volume\&. The label must already be in the tapelist file\&.
.RE
.PP
\fB\-f\fR
.RS 4
Force the label operation; see above
.RE
.PP
\fB\-\-version\fR
.RS 4
Output version information
.RE
.PP
\fB\-o \fR\fB\fIconfigoption\fR\fR
.RS 4
See the "CONFIGURATION OVERRIDE" section in
\fBamanda\fR(8)\&.
.RE
.SH "EXAMPLE"
.PP
Write an Amanda label with the string "DMP000" on the current volume:
.nf
% amlabel daily DMP000 
.fi
.PP
Label the tape in slot 3 of the changer with the label "DMP003":
.nf
% amlabel daily DMP003 slot 3 
.fi
.SH "MESSAGES"
.PP
Label \*(Aq\fIlabel\fR\*(Aq doesn\*(Aqt match labelstr \*(Aq\fIlabelstr\fR\*(Aq
.RS 4
The given label does not match the configured labelstr\&. Even with
\fB\-f\fR, this is an error\&.
.RE
.PP
Label \*(Aq\fIlabel\fR\*(Aq already on a volume
.RS 4
The given label is already on another volume in the tapelist, and Amanda will not write it to this volume unless forced\&.
.RE
.PP
Found label \*(Aq\fIlabel\fR\*(Aq, but it is not from configuration \fIconfig\fR\&.
.RS 4
The label
\fBamlabel\fR
found on the volume does not match this configuration\*(Aqs
\fBlabelstr\fR, so it is probably part of a different Amanda configuration\&. Amanda will not overwrite the label unless it is forced\&.
.RE
.PP
Volume with label \*(Aq\fIlabel\fR\*(Aq contains data from this configuration\&.
.RS 4
The label
\fBamlabel\fR
found on the volume still has valid data on it and Amanda will not overwrite it unless forced\&. Note that this does not distinguish between active and inactive volumes\&.
.RE
.PP
.RS 4
.RE
.SH "SEE ALSO"
.PP
\fBamanda\fR(8),
\fBamanda.conf\fR(5),
\fBtapelist\fR(5),
\fBamanda-taperscan\fR(7),
\fBamdump\fR(8),
\fBamflush\fR(8)
.PP
The Amanda Wiki:
: http://wiki.zmanda.com/
.SH "AUTHORS"
.PP
\fBJames da Silva\fR <\&jds@amanda\&.org\&>
.PP
\fBStefan G\&. Weichinger\fR <\&sgw@amanda\&.org\&>
.PP
\fBDustin J\&. Mitchell\fR <\&dustin@zmanda\&.com\&>
.RS 4
Zmanda, Inc\&. (http://www\&.zmanda\&.com)
.RE


NEWS
Changes in release 2.5.0p2

* Fix major bug in amandad.
* amverifyeun works.
* amstatus works with ':' in diskname.

Changes in release 2.5.0p1

* Add the 'amtape update' command.
* Many small bug fix.

Changes in release 2.5.0

* Communication security/authentication: Kerberos 4/5, OpenSSH
* Data security: Symmetric/Assymetric encrytion algorithms (aesutil
  and gpg encryption),  Encryption can be done on server or client,
  Custom encryption utilities can be used.
* Compression: Ability to add custom compression tools. This is a
  really useful feature especially since it can specified for a DLE.  You
  can use different compression algorithm for images, binaries, ascii files
  and so on.
* Dump images spanning multiple media volumes - Dump images are no
  longer restricted to a single media volume (tape or vtape).  Data restoration
  can be done using amrecover and amfetchdump commands.
* Auto tape labelling - This optional feature is good for disk backups.

Change in release 2.4.5p1

* New [--days <num>] argument to 'amadmin <conf> tape'.
* New amoverview -last -num0 -togo0 options.

Change in release 2.4.5

* new displayunit global option to select the unit use to display number
  k=kilo, m=mega, g=giga, t=tera.
* new amoverview -skipmissed option.

Change in release 2.4.5b1

* holding disk disk use timestamped directory.
* autoflush flush today's dump.
* new bumppercent global option, this should improve bumping criteria,
  the bumpsize is set to a fixed value which can be adequate for small
  and large disk.
* bumpsize, bumppercent, bumpdays and bumpmult can be in a dumptype.
* calcsize support include and exclude like gnutar.
* new 'estimate' dumptype option to select estimate type:
    CLIENT: estimate by the dumping program.
    CALCSIZE: estimate by the calcsize program, a lot faster but less acurate.
    SERVER: estimate based on statistic from previous run, take second but
            can be wrong on the estimate size.

Changes in release 2.4.4p3

* new chg-iomega changer script.
* amanda will not use a tape if it's label is not in the tapelist file
  (eg. after an amrmtape).

Changes in release 2.4.4p2

* new initial_poll_delay config in chg-zd-mtx
* 'amadmin find' list disk removed from the disklist.
* amrecover can recover a disk removed from the disklist file.
* amanda works with samba3.
* new chg-disk changer script to use with the file: driver.
* Kerberos 4 support integrated and functional again.

Changes in release 2.4.4p1

* amverify do not advance to next tape at the end.
* new amflush -b and -s options, change -f behavior.
* new chg-mcutil changer.
* amrecover_changer works with chg-multi.
* default driveslot for chg-zd-mtx is now 0.
* amplot generate color postscript by default.
* amplot -b flag to generate b/w postscript.
* rait: works with file:
* print a postscript label for each tape used.

Changes in release 2.4.4

* New -a option to amcheck to always send an email.

Changes in release 2.4.4b1

* maxpromoteday: New option for a dumptype.
* New amtapetype program (replace tapetype)
* Client compile on CYGWIN
* amrestore -f <fileno> : amrestore can do fsf before reading the tape
* amrestore -l <label>  : amrestore can check the label before restoring.
* New config options:
  amrecover_do_fsf: Amrecover will use the -f flag of amrestore to position the
                    tape.
  amrecover_check_label: amrecover will pass the -l flag to amrestore.
  amrecover_changer: amrecover will use your changer.
  maxdumpsize: The maximum size of the dumps during a run,
               default to tapesize * runtapes
  taperalgo: select your algo for the taper, try 'largestfit', it can improve
             tape usage.
* amrecover allow to retry a skip a tape.
* New --days option to 'amadmin <config> balance' command.
* New --date option to amstatus
* amreport print an usage by tape.
* Newer promote algorithm that try to reduce the number of full on a
  single host in a run.
* New changer: chg-juke, chg-null and chg-rait

Changes in release 2.4.3

* Few small bug fixed
* New runtapes argument to amverify.
* New amverifyrun program.
* 2.4.3 client works with older server.

Changes in release 2.4.3b4

* new noop packet type. It allow the server to know the client features.
* exclude list with relative path is working again.
* amstatus is working again.

Changes in release 2.4.3b3

* --with-maxtapeblocksize configure options
* blocksize tapetype option
* file-pad tapetype option
* Multiple exclude in dumptype
* Option include in dumptype
* New disklist syntax:
*   hostname diskname [ diskdevice ] dumptype [ spindle [ interface ] ]
* chg-zd-mtx: Major cleanup and general overhaul.
* amrecover: new listdisk command.

Changes in release 2.4.3b2

* amrestore try to find the next chunk in the current directory.
* amrecover: You can change the tape device between each tape.
* amstatus work with amflush.

Changes in release 2.4.3b1

* tapeio.
* New [host [disk]*]* arguments to amdump, amcheck, amadmin and amflush.
* New [-D datastamp]* arguments to amflush.
* amrecover: cd accept shell wildcard, new command cdx accept regex.
* new autoflush option.
* new dumperstr option to specify the priority order of each dumper.

Changes in release 2.4.2

* Samba passwords are now sent to smbclient via a pipe and never displayed.
* Debug files in /tmp/amanda (--with-debug-dir) are now timestamped and old
  ones automatically cleaned out.  This means more space (a few KBytes) will
  be used since in a given run, several of the programs are called more than
  once.  But it also means important debugging information should no longer
  be lost by the file being overwritten.  The length of time to keep the
  files is controlled by --with-debug-days (default: 4).  The old flag
  --with-pid-debug-files is no longer needed and is ignored.
* dumptypes can be specialized in the disklist file
* Amanda will never use more space than specified (use) for an holding
  disk

server-src/amlabel.c
/*
 * Amanda, The Advanced Maryland Automatic Network Disk Archiver
 * Copyright (c) 1991-1998, 2000 University of Maryland at College Park
 * All Rights Reserved.
 *
 * Permission to use, copy, modify, distribute, and sell this software and its
 * documentation for any purpose is hereby granted without fee, provided that
 * the above copyright notice appear in all copies and that both that
 * copyright notice and this permission notice appear in supporting
 * documentation, and that the name of U.M. not be used in advertising or
 * publicity pertaining to distribution of the software without specific,
 * written prior permission.  U.M. makes no representations about the
 * suitability of this software for any purpose.  It is provided "as is"
 * without express or implied warranty.
 *
 * U.M. DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING ALL
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS, IN NO EVENT SHALL U.M.
 * BE LIABLE FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION
 * OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
 * CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 *
 * Authors: the Amanda Development Team.  Its members are listed in a
 * file named AUTHORS, in the root directory of this distribution.
 */
/*
 * $Id: amlabel.c,v 1.53 2006/07/25 18:27:57 martinea Exp $
 *
 * write an Amanda label on a tape
 */
#include "amanda.h"
#include "conffile.h"
#include "tapefile.h"
#include "changer.h"
#include <device.h>
#include <timestamp.h>
#include <taperscan.h>

/* local functions */

int main(int argc, char **argv);

static void usage(void) {
    g_fprintf(stderr, _("Usage: %s [-f] <conf> <label> [slot <slot-number>] [-o configoption]*\n"),
	    get_pname());
    exit(1);
}

static void print_read_label_status_error(ReadLabelStatusFlags status) {
    char ** status_strv;

    if (status == READ_LABEL_STATUS_SUCCESS)
        return;

    status_strv = g_flags_nick_to_strv(status,
                                       READ_LABEL_STATUS_FLAGS_TYPE);
    g_assert(g_strv_length(status_strv) > 0);
    if (g_strv_length(status_strv) == 1) {
        g_printf("Error was %s.\n", *status_strv);
    } else {
        char * status_list = g_english_strjoinv(status_strv, "or");
        g_printf("Error was one of %s.\n", status_list);
        amfree(status_list);
    }
    g_strfreev(status_strv);
}

int
main(
    int		argc,
    char **	argv)
{
    char *conf_tapelist;
    char *outslot = NULL;
    char *label, *tapename = NULL;
    char *labelstr, *slotstr;
    char *conf_tapelist_old;
    int have_changer;
    int force, tape_ok;
    tapetype_t *tape;
    size_t tt_blocksize_kb;
    int slotcommand;
    Device * device;
    ReadLabelStatusFlags label_status;
    char *cfg_opt = NULL;
    config_overwrites_t *cfg_ovr = NULL;

    /*
     * Configure program for internationalization:
     *   1) Only set the message locale for now.
     *   2) Set textdomain for all amanda related programs to "amanda"
     *      We don't want to be forced to support dozens of message catalogs.
     */  
    setlocale(LC_MESSAGES, "C");
    textdomain("amanda"); 

    safe_fd(-1, 0);
    safe_cd();

    set_pname("amlabel");

    dbopen(DBG_SUBDIR_SERVER);
    device_api_init();

    /* Don't die when child closes pipe */
    signal(SIGPIPE, SIG_IGN);

    malloc_size_1 = malloc_inuse(&malloc_hist_1);

    erroutput_type = ERR_INTERACTIVE;

    cfg_ovr = extract_commandline_config_overwrites(&argc, &argv);

    if(argc > 1 && strcmp(argv[1],"-f") == 0)
	 force=1;
    else force=0;

    if(argc != 3+force && argc != 5+force)
	usage();

    cfg_opt = argv[1+force];
    label = argv[2+force];

    if(argc == 5+force) {
	if(strcmp(argv[3+force], "slot"))
	    usage();
	slotstr = argv[4+force];
	slotcommand = 1;
    } else {
	slotstr = "current";
	slotcommand = 0;
    }

    config_init(CONFIG_INIT_EXPLICIT_NAME | CONFIG_INIT_FATAL,
		cfg_opt);
    apply_config_overwrites(cfg_ovr);

    check_running_as(RUNNING_AS_DUMPUSER);

    dbrename(config_name, DBG_SUBDIR_SERVER);

    conf_tapelist = config_dir_relative(getconf_str(CNF_TAPELIST));
    if (read_tapelist(conf_tapelist)) {
	error(_("could not load tapelist \"%s\""), conf_tapelist);
	/*NOTREACHED*/
    }

    labelstr = getconf_str(CNF_LABELSTR);

    if(!match(labelstr, label)) {
	error(_("label %s doesn't match labelstr \"%s\""), label, labelstr);
	/*NOTREACHED*/
    }

    if((lookup_tapelabel(label))!=NULL) {
	if(!force) {
	    error(_("label %s already on a tape\n"),label);
	    /*NOTREACHED*/
    	}
    }
    tape = lookup_tapetype(getconf_str(CNF_TAPETYPE));
    tt_blocksize_kb = (size_t)tapetype_get_blocksize(tape);

    if((have_changer = changer_init()) == 0) {
	if(slotcommand) {
	    g_fprintf(stderr,
	     _("%s: no tpchanger specified in \"%s\", so slot command invalid\n"),
		    argv[0], config_filename);
	    usage();
	}
	tapename = getconf_str(CNF_TAPEDEV);
	if (tapename == NULL) {
	    error(_("No tapedev specified"));
	}
    } else if(have_changer != 1) {
	error(_("changer initialization failed: %s"), strerror(errno));
	/*NOTREACHED*/
    } else {
	if(changer_loadslot(slotstr, &outslot, &tapename)) {
	    error(_("could not load slot \"%s\": %s"), slotstr, changer_resultstr);
	    /*NOTREACHED*/
	}
    }
#endif /* HAVE_LINUX_ZFTAPE_H */

	g_printf(_("labeling tape in slot %s (%s):\n"), outslot, tapename);
    }

    tape_ok=1;
    g_printf("Reading label...\n");fflush(stdout);
    device = device_open(tapename);
    if (device == NULL) {
        error("Could not open device %s.\n", tapename);
    }
    
    device_set_startup_properties_from_config(device);
    label_status = device_read_label(device);

    if (label_status & READ_LABEL_STATUS_VOLUME_UNLABELED) {
        g_printf("Found an unlabeled tape.\n");
    } else if (label_status != READ_LABEL_STATUS_SUCCESS) {
        g_printf("Reading the tape label failed: \n  ");
        print_read_label_status_error(label_status);
        tape_ok = 0;
    } else {
	/* got an amanda tape */
	g_printf(_("Found Amanda tape %s"),device->volume_label);
	if(match(labelstr, device->volume_label) == 0) {
	    g_printf(_(", but it is not from configuration %s."), config_name);
	    if(!force)
		tape_ok=0;
	} else {
	    if((lookup_tapelabel(device->volume_label)) != NULL) {
		g_printf(_(", tape is active"));
		if(!force)
		    tape_ok=0;
	    }
	}
	g_printf("\n");
    }

    if(tape_ok) {
	char *timestamp = NULL;

	g_printf(_("Writing label %s..\n"), label); fflush(stdout);
        
	timestamp = get_undef_timestamp();
        if (!device_start(device, ACCESS_WRITE, label, timestamp)) {
	    error(_("Error writing label.\n"));
            g_assert_not_reached();
	} else if (!device_finish(device)) {
            error(_("Error closing device.\n"));
            g_assert_not_reached();
        }
	amfree(timestamp);

        g_printf(_("Checking label...\n")); fflush(stdout);

        label_status = device_read_label(device);
        if (label_status != READ_LABEL_STATUS_SUCCESS) {
            g_printf("Checking the tape label failed: \n  ");
            print_read_label_status_error(label_status);
            exit(EXIT_FAILURE);
        } else if (device->volume_label == NULL) {
            error(_("no label found.\n"));
            g_assert_not_reached();
        } else if (strcmp(device->volume_label, label) != 0) {
            error(_("Read back a different label: Got %s, but expected %s\n"),
                  device->volume_label, label);
            g_assert_not_reached();
        } else if (get_timestamp_state(device->volume_time) !=
                   TIME_STATE_UNDEF) {
            error(_("Read the right label, but the wrong timestamp: "
                    "Got %s, expected X.\n"), device->volume_time);
            g_assert_not_reached();
        }
        
        /* write tape list */
        
        /* make a copy */
        conf_tapelist_old = stralloc2(conf_tapelist, ".amlabel");
        if(write_tapelist(conf_tapelist_old)) {
            error(_("couldn't write tapelist: %s"), strerror(errno));
            /*NOTREACHED*/
        }
        amfree(conf_tapelist_old);
        
        /* XXX add cur_tape number to tape list structure */
        remove_tapelabel(label);
        add_tapelabel("0", label);
        if(write_tapelist(conf_tapelist)) {
            error(_("couldn't write tapelist: %s"), strerror(errno));
            /*NOTREACHED*/
        }
        
        g_printf(_("Success!\n"));
    } else {
	g_printf(_("\ntape not labeled\n"));
    }
    
    g_object_unref(device);
    device = NULL;

    clear_tapelist();
    amfree(outslot);
    amfree(conf_tapelist);
    config_name=NULL;
    dbclose();

    return 0;
}
